Doc Blocks: Fix styling and parameter bugs

JReinhold opened the pull request:
Closes #20850 

<!-- Thank you for contributing to Storybook! If your PR is related to an issue, provide the number(s) above, e.g. #1000, #1001 -->

## What I did

This PR fixes a few outstanding bugs with docs and blocks.

1. Canvases in autodocs didn't respect `parameters.docs.canvas.withToolbar = false` - they do now.
2. We introduced a regression that broke Canvas when used in unattached mode and without any story children (deprecated API). This has now been fixed and has an internal story. This was reported here https://github.com/storybookjs/storybook/issues/19964#issuecomment-1405161473
4. inheritable styles like fonts were set on the docs content root, which would bleed through Unstyled blocks. We no longer set that reset, and it was a mistake it got set in the first place. This ensures that Unstyled removes styles from fonts.
5. Headers were not unstyled when wrapped in Unstyled. They are now. This happened because we still replace headers in MDX with custom components, and they had local styling. Now that local styling is removed so they rely solely on the global styles, which Unstyled can disable.

There's still a difference in header styles between autodocs and manual docs, specifically I've noticed that `h2` in autodocs have font-weight 400, while they have font-weight 700 in manual docs. font-weight 400 is a global style we set at the manager level via Emotion, and I can't figure out why that global style is not set in manual docs. It's not because it's overwritten, it is never there.
If anyone has inputs to this let me know.


<!-- Briefly describe what your PR does -->

## How to test

- See new internal story for Canvas in UI SB
- See the Unstyled docs page in UI SB
- See new stories for Stories and Primary blocks in UI SB

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

JReinhold committed:
fix canvas.withToolbar parameter in autodocs

JReinhold committed:
make headers in MDX unstyled components

JReinhold committed:
improve withToolbar logic

JReinhold committed:
add stories for Primary and Stories

JReinhold committed:
Support unattached Canvas without story children

JReinhold committed:
fix type error

JReinhold committed:
Merge branch 'next' into jeppe/fix-blocks

renato-bohler commented:
Does this fix https://github.com/storybookjs/storybook/issues/20780?

tmeasday commented on code/ui/blocks/src/blocks/DocsStory.tsx:
```diff
@@ -10,12 +10,17 @@ import { useOf } from './useOf';
   const { story } = useOf(of || 'story', ['story']);
 
+  // use withToolbar from parameters or default to true in autodocs
+  const withToolbar =
+    story.parameters.docs?.canvas?.withToolbar !== undefined
+      ? story.parameters.docs.canvas.withToolbar
+      : true;
```
This change means the `Stories` block gets `withToolbar` by default which is a regression:

<img width="903" alt="image" src="https://user-images.githubusercontent.com/132554/214984242-cf3e4d40-8cd3-4645-bb4a-59e3a038ba34.png">

JReinhold replied:
oh yeah, that was how my mental model for the autodocs were, but now that I think about it, you're right that only the controllable story gets the toolbar. I'll fix that.

tmeasday commented:
@JReinhold I see font-weight 700 on h2s in autodocs?

<img width="2255" alt="image" src="https://user-images.githubusercontent.com/132554/214984762-4c06d528-88dc-422b-b5df-60514f7bea3d.png">

JReinhold commented:
> @JReinhold I see font-weight 700 on h2s in autodocs?
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/132554/214984762-4c06d528-88dc-422b-b5df-60514f7bea3d.png">

Uh... I don't? Neither on next nor in this branch. This is from the published SB in this branch.
What's going on? It's even the same class hash.

![image](https://user-images.githubusercontent.com/5678122/215039486-a679ef96-15ef-4bfb-b735-f1d9af44845f.png)

JReinhold committed:
improve withToolbar logic

Co-authored-by: Tom Coleman <tom@chromatic.com>

JReinhold commented:
> Would that fix #20780?

Yes, but that was also fixed by #20780 so no need to here: 👍

JReinhold committed:
fix toolbar appearing in Stories

JReinhold committed:
fix __forceInitialArgs to not update source

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/fix-blocks

JReinhold commented:
@tmeasday I've added a commit that fixes #20745, and reacted to your feedback, could you re-review?

in essence I added `initalCode` to the `SourceContext`, that will always be the first snippet rendered. when Canvas gets `forceInitial` it will make sure to only use `initialCode` and not `code`. This breaks down if any of our frameworks for some reason emits different sources initially, as we'd "stop" at the first one. But I don't know if that is a scenario that exists.

(I've changed `story.__forceInitialArgs` prop on the Canvas only to `__forceInitial` as it now affects both the args and the code)

If we wanted a more proper solution for this we should probably stop rendering the same story twice in autodocs and instead generate a virtual story that we'd reference to ensure they don't collide. But I don't think now is the time for that.

tmeasday commented:
@JReinhold the rest of the changes look good, but I don't think https://github.com/storybookjs/storybook/pull/20803/commits/16eb6966d2b0769fcac89fc80a9a1f1bd0712eec is the right fix here.

It would be an issue if you

a) changed the args
b) visited the docs page

Then the primary story would render & emit with the updated args, and the second render would use + emit the story's `initialArgs`. Which one would reach the source container first... I'm not sure it's defined, but probably the primary :)

Maybe pull that commit into a second PR and let's discuss further there?

The solution suggested by @shilman was to somehow add an extra description prop to the `Story` + `Source` blocks to tie them together (that prop's value would get passed to the `SourceContainer` somehow). So we'd render like:

```mdx
<Story of={...} sourceLabel="primary" />
<Source of={...} sourceLabel="primary" />
```

WDYT about that? (don't love the name `sourceLabel`, and not sure if it should be internal or not).

JReinhold commented:
> @JReinhold the rest of the changes look good, but I don't think [16eb696](https://github.com/storybookjs/storybook/commit/16eb6966d2b0769fcac89fc80a9a1f1bd0712eec) is the right fix here.
> 
> It would be an issue if you
> 
> a) changed the args b) visited the docs page

You're right, I missed that.

> Maybe pull that commit into a second PR and let's discuss further there?

Deal!

> The solution suggested by @shilman was to somehow add an extra description prop to the `Story` + `Source` blocks to tie them together (that prop's value would get passed up to the `SourceContainer` somehow). So we'd render like:
> 
> ```gfm
> <Story of={...} sourceLabel="primary" />
> <Source of={...} sourceLabel="primary" />
> ```
> 
> WDYT about that? (don't love the name `sourceLabel`, and not sure if it should be internal or not).

So sort of like having a "sub story id" to key and get sources by, I think that makes sense. I don't know if it should be hardcoded by the rendering components, or if we should just generate a uuid to use instead.

JReinhold committed:
Revert "fix __forceInitialArgs to not update source"

This reverts commit 16eb6966d2b0769fcac89fc80a9a1f1bd0712eec.

JReinhold committed:
Merge branch 'next' into jeppe/fix-blocks

tmeasday commented:
> or if we should just generate a uuid to use instead.

Are you imagining each `Canvas` would generate its own `uuid`? I guess that could work, but would have to be stable across re-renders.

I'm also wondering if (one day) we might also allow more than one set of arg values for the same story. We could use a similar "sub story id" concept to disambiguate the events. So the `Controls` block would need to know about the sub story id also.

JReinhold committed:
allow Canvas without story children

JReinhold committed:
Merge branch 'jeppe/fix-blocks' of github.com:storybookjs/storybook into jeppe/fix-blocks

JReinhold commented on code/ui/blocks/src/blocks/Source.tsx:
```diff
@@ -118,8 +118,7 @@ export const useSourceProps = (
       stories = [docsContext.storyById()];
     } catch (err) {
-      // You are allowed to use <Story code="..." /> unattached.
-      if (!props.code) throw err;
```
@tmeasday I had to remove this check to make story-less, unattached Canvases work.

JReinhold replied:
Fails gracefully 
Source:
![image](https://user-images.githubusercontent.com/5678122/215441225-4cc2a585-1486-4013-b8ff-cc6553dcb09f.png)
Canvas:
![image](https://user-images.githubusercontent.com/5678122/215441337-f88a9023-ce67-4e82-be13-1e9af069bfcb.png)

JReinhold committed:
fix addon docs e2e test

JReinhold committed:
add Empty Unattached story to Source block

JReinhold committed:
add docs styles to block stories

JReinhold commented:
@tmeasday I changed all stories for blocks to be decorated by the `DocsContainer` - using `parameters.docsStyles = true`. I kept seeing some inconsistencies between docs view and story view for the blocks, and hopefully this should make it more consistent, as they now inherit the same global styles as in docs view.

This does add some unnecessary margin to all the stories, but I think that's a fine tradeoff. pixels are free.
